L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTSAMuonProducer.cc

```diff
@@ -0,0 +1,196 @@
+
+  // ----------member data ---------------------------
+  edm::EDGetTokenT<BXVector<l1t::Muon> > muonToken_;
+  unsigned int Nprompt;
+  unsigned int Ndisplaced;
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+Phase2L1TGMTSAMuonProducer::Phase2L1TGMTSAMuonProducer(const edm::ParameterSet& iConfig)
+    : muonToken_(consumes<l1t::MuonBxCollection>(iConfig.getUntrackedParameter<edm::InputTag>("muonToken"))),
+      Nprompt(iConfig.getParameter<uint>("Nprompt")),
+      Ndisplaced(iConfig.getParameter<uint>("Ndisplaced"))
+{
+  produces<std::vector<l1t::SAMuon> >("promptSAMuons").setBranchAlias("prompt");
+  produces<std::vector<l1t::SAMuon> >("displacedSAMuons").setBranchAlias("displaced");
+}
+
+Phase2L1TGMTSAMuonProducer::~Phase2L1TGMTSAMuonProducer() {}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void Phase2L1TGMTSAMuonProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  edm::Handle<l1t::MuonBxCollection> muon;
+  iEvent.getByToken(muonToken_, muon);
+
+  // Output
+  std::vector<SAMuon> prompt;
+  std::vector<SAMuon> displaced;
+
+  for (int bx = muon->getFirstBX(); bx <= muon->getLastBX(); ++bx) {
+    //TODO: We are expecting to send all BX. Using bx0 for now
+    if (bx != 0) {
+      continue;
+    }
+
+    for (uint i = 0; i < muon->size(bx); ++i) {
+      const l1t::Muon& mu = muon->at(bx, i);
+
+      //TODO: Still looking for a way to get displaced muon
+      if (abs(mu.hwDXY()) > 0)
+        displaced.push_back(Convertl1tMuon(mu, bx));
+      else
+        prompt.push_back(Convertl1tMuon(mu, bx));
+    }
+
+    // Sort by hwPt
+    std::sort(prompt.begin(), prompt.end(), std::greater<>());
+    std::sort(displaced.begin(), displaced.end(), std::greater<>());
+
+    // Store into output, allow up to 18 prompt + 18 displayed
+    if (prompt.size() > Nprompt) {
+      prompt.resize(Nprompt);
+    }
+    if (displaced.size() > Ndisplaced) {
+      displaced.resize(Ndisplaced);
+    }
+  }
+
+  std::unique_ptr<std::vector<l1t::SAMuon> > prompt_ptr = std::make_unique<std::vector<l1t::SAMuon> >(prompt);
+  std::unique_ptr<std::vector<l1t::SAMuon> > displaced_ptr = std::make_unique<std::vector<l1t::SAMuon> >(displaced);
+  iEvent.put(std::move(prompt_ptr), "promptSAMuons");
+  iEvent.put(std::move(displaced_ptr), "displacedSAMuons");
+}
+
+// ===  FUNCTION  ============================================================
+//         Name:  Phase2L1TGMTSAMuonProducer::Convertl1tMuon
+//  Description:
+// ===========================================================================
+SAMuon Phase2L1TGMTSAMuonProducer::Convertl1tMuon(const l1t::Muon& mu, const int bx_) {
+  ap_uint<BITSQUALITY> qual = mu.hwQual();
+  // We might not need to send BX info
+  //ap_int<BITSBX> bx(bx_);
+  bool charge(mu.charge() > 0);
```

Comment:
Also this is inconsistent with the [interface document](https://www.overleaf.com/read/ptkfbjyyxcvv), that has the charge bit defined as '0 = positive, 1 = negative'.
I would be happy with this new definition, as it would match with what we're using in the correlator, but you should at least inform us when you're changing things, as we need to propagate those changes to our firmware.